src/vm/diagnosticserver.cpp

```diff
@@ -199,11 +201,35 @@ bool DiagnosticServer::Shutdown()
                 STRESS_LOG2(
                     LF_DIAGNOSTICS_PORT,                                  // facility
                     LL_ERROR,                                             // level
-                    "Failed to unlink diagnostic IPC: error (%d): %s.\n", // msg
+                    "Failed to close diagnostic IPC: error (%d): %s.\n",  // msg
                     code,                                                 // data1
                     szMessage);                                           // data2
             };
-            s_pIpc->Unlink(ErrorCallback);
+            s_pIpc->Close(ErrorCallback); // This will break the accept waiting for client connection.
+
+            if (s_hServerThread != NULL)
+            {
+#ifndef FEATURE_PAL
+                if (::CancelSynchronousIo(s_hServerThread) == 0)
+                {
+                    _ASSERTE(!"Failed to mark pending synchronous I/O operations issued by Diagnostics Server Thread as canceled.");
```

Comment:
Did you want to remove this assert now that we know it is problematic?

Reply:
Just cherry-picked it from merged PR on master (easier to track history).